core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,292 @@
+    }
+    XMLInputFormat.addInputPaths(job, config.path);
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_REPROCESSING_REQUIRED, this.config.reprocessingRequired);
+    //set file tracking information in a temporary file, to be available to read outside plugin.
+    setFileTrackingInfo(context, conf);
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION, this.config.actionAfterProcess);
+    }
+    if (StringUtils.isNotEmpty(this.config.targetFolder)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, this.config.targetFolder);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, this.config.nodePath);
+    conf.set(XMLInputFormat.XML_INPUT_NAME_CONFIG, config.path);
+    XMLInputFormat.setInputPathFilter(job, BatchXMLFileFilter.class);
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(XMLInputFormat.class, conf)));
+  }
+
+  /**
+   * Method to set file tracking information.
+   */
+  private void setFileTrackingInfo(BatchSourceContext context, Configuration conf) {
+    String tableName = this.config.tableName;
+    processedFileTrackingTable = context.getDataset(tableName);
+    if (processedFileTrackingTable != null) {
+      List<String> processedFiles = new ArrayList<String>();
+      Calendar cal = Calendar.getInstance();
+      cal.add(Calendar.DATE, -Integer.valueOf(config.tableExpiryPeriod));
+      Date expiryDate = cal.getTime();
+
+      CloseableIterator<KeyValue<byte[], byte[]>> iterator = processedFileTrackingTable.scan(null, null);
+      while (iterator.hasNext()) {
+        KeyValue<byte[], byte[]> keyValue = iterator.next();
+        //delete record before expiry time period
+        Long time = new Long(new String(keyValue.getValue(), Charsets.UTF_8));
+        Date processedDate = new Date(time);
+        if (processedDate.before(expiryDate)) {
+          processedFileTrackingTable.delete(keyValue.getKey());
+        } else {
+          processedFiles.add(new String(keyValue.getKey(), Charsets.UTF_8));
+        }
+      }
+      //TODO - remove temp file name setting in config after proper solution.
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE, tableName);
+      //For reprocessing not required, set processed file name to configuration.
+      //File name use by BatchXMLFileFilter to filter already processed files.
+      if (config.reprocessingRequired.equalsIgnoreCase("NO")) {
+        conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_FILES,
+                 GSON.toJson(processedFiles, ARRAYLIST_PREPROCESSED_FILES));
+      }
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRecord> emitter) throws Exception {
+    Map<String, String> xmlRecord = (Map<String, String>) input.getValue();
+    Set<String> keySet = xmlRecord.keySet();
+    Iterator<String>  itr = keySet.iterator();
```

Comment:
Use `Iterators.getOnlyElement`, instead of `itr.next()`. This'll enforce the expectation that there's only one element in the iterator.